File: Data Structures & Algorithms/subsets/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums) {
        let r = new Array();
        let s = new Array();
        this.dfs(nums, 0, s, r);
        return r;
    }

    dfs(n, i, s, r) {
        if (i >= n.length) {
            r.push([...s]);
            return;
        }

        s.push(n[i]);
        this.dfs(n, i+1, s, r);

        s.pop();
        this.dfs(n, i+1, s, r);
    }
}

File: Data Structures & Algorithms/validate-parentheses/submission-0.py
class Solution:
    def isValid(self, s: str) -> bool:
        mapping = {
            ")": "(",
            "]": "[",
            "}": "{"
        }
        stack = []

        for p in s:
            if p in mapping:
                if stack and stack[-1] == mapping[p]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(p)
        
        return True if not stack else False
